chemprop/models/ffn.py

```diff
@@ -0,0 +1,350 @@
+                num_layers=weights_ffn_num_layers,
+                dropout=dropout,
+                activation=activation,
+            )
+        else:
+            if shared_ffn:
+                self.ffn_readout = nn.Sequential(
+                    ffn_base,
+                    build_ffn(
+                        first_linear_dim=base_output_size,
+                        hidden_size=hidden_size,
+                        num_layers=1,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                )
+            else:
+                self.ffn_readout = build_ffn(
+                    first_linear_dim=features_size,
+                    hidden_size=hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                )
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.Tensor, List, torch.Tensor, List, torch.Tensor],
+        constraints: torch.Tensor,
+        bond_types: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise ValueError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if constraints is not None:
+            if self.ffn_type == "atom":
+                output_hidden = self.ffn(hidden)
+                output = self.ffn_readout(output_hidden)
+            elif self.ffn_type == "bond":
+                b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+                b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+                output_1 = self.ffn(b_hidden_1)
+                output_2 = self.ffn(b_hidden_2)
+                output_hidden = (output_1 + output_2) / 2
+                output = self.ffn_readout(output_hidden)
+                if bond_types is not None:
+                    output = output + bond_types.reshape(-1, 1)
+
+            u_a = self.weights_readout(output_hidden)
```

Comment:
```suggestion
            W_a = self.weights_readout(output_hidden)
```
the formula in the paper uses `W`

Reply:
This is the $u\hat{a_i}$ in equation 14 in the SI. It would become ${w_i}$ after being passed in the softmax function.

Reply:
why even bother with it then? `u_a` here is only used in the context of being passed into the `Softmax` function and the term only appears in the SI. If we can avoid terms from the SI in favor of only using terms from the main text (as we can here), we should do so